base/abstractarray.jl

```diff
@@ -86,8 +128,48 @@ linearindexing(A::AbstractArray, B::AbstractArray...) = linearindexing(linearind
 linearindexing(::LinearFast, ::LinearFast) = LinearFast()
 linearindexing(::LinearIndexing, ::LinearIndexing) = LinearSlow()
 
+abstract IndicesBehavior
+immutable IndicesStartAt1  <: IndicesBehavior end   # indices 1:size(A,d)
+immutable IndicesUnitRange <: IndicesBehavior end   # arb UnitRange indices
+immutable IndicesList      <: IndicesBehavior end   # indices like (:cat, :dog, :mouse)
+
+indicesbehavior(A::AbstractArray) = indicesbehavior(typeof(A))
+indicesbehavior{T<:AbstractArray}(::Type{T}) = IndicesStartAt1()
+indicesbehavior(::Number) = IndicesStartAt1()
+
+abstract IndicesPerformance
+immutable IndicesFast1D <: IndicesPerformance end  # indices(A, d) is fast
+immutable IndicesSlow1D <: IndicesPerformance end  # indices(A) is better than indices(A,d)
+
+indicesperformance(A::AbstractArray) = indicesperformance(typeof(A))
+indicesperformance{T<:AbstractArray}(::Type{T}) = IndicesFast1D()
+
+"""
+    shape(A)
+
+Returns a tuple specifying the "shape" of array `A`. For arrays with
+conventional indexing (indices start at 1), this is equivalent to
+`size(A)`; otherwise it is equivalent to `incides(A)`.
+"""
```

Comment:
Ah, I understand better now. I still think it's somewhat problematic, especially since numbers are still iterable. It might be better to have `indices` of an `Array` return something like `(UpTo(3),UpTo(5))` to clearly distinguish sizes from index ranges.